### ISSUE TITLE
Mypy doesn't behave well with typing.overload-like decorators

### ISSUE THREAD

**Issue by MiguelGuthridge** [python/mypy#12589]
**Bug Report**

I'm trying to create a library where a decorator can be used to manage multiple definitions of a function or class. Much like the `typing.overload` decorator, I only want the final definition of the function to be used by Python.

**To Reproduce**

This is some (heavily simplified) code that I'm using to implement this functionality

```py
def forget(func) -> None:
    # In reality I'm doing a little more than just returning nothing
    return None

# This version effectively doesn't exist after it is decorated
@forget
def hello():
    pass
# Therefore defining this function shouldn't cause an error
# Name "hello" already defined on line 10
def hello():
    pass
```

If I'm annotating this incorrectly, I'd appreciate some info on the correct method, since I couldn't find any documentation on it.

**Expected Behavior**

I'd expect when I run this code through mypy, I won't get any issues, as I'm replacing the function with nothing.

**Actual Behavior**

I get the error documented in the comments.

**Your Environment**

- Mypy version used: 0.942
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): None
- Python version used: 3.10.0
- Operating system and version: Kubuntu 21.10

**Comment by AlexWaygood** [comment]
Basically the same issue as #11065 -- would be best to leave a comment there instead.

Mypy doesn't have a great deal of support for this kind of pattern at the moment, unfortunately :/

**Comment by MiguelGuthridge** [comment]
Sadly I don't think that the proposed solution will work very well as it would require users of my library to specify the final function's name while decorating every single function, which would be come very convoluted and difficult to maintain.

**Comment by AlexWaygood** [comment]
> Sadly I don't think that the proposed solution will work very well as it would require users of my library to specify the final function's name while decorating every single function, which would be come very convoluted and difficult to maintain.

Yes, I know. I agree that the current situation is far from ideal. It would be good if mypy had better support for this pattern, which is why https://github.com/python/mypy/issues/11065 is still an open issue.

Please leave a comment at https://github.com/python/mypy/issues/11065 with your concerns :)

**Comment by MiguelGuthridge** [comment]
Thanks for the clarification, I didn't spot that it was still open.

**Comment by AlexWaygood** [comment]
> Thanks for the clarification, I didn't spot that it was still open.

No problem :)
